"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportRevertedWithPanic = void 0;
const common_1 = require("hardhat/common");
const panic_1 = require("./panic");
const utils_1 = require("./utils");
function supportRevertedWithPanic(Assertion) {
    Assertion.addMethod("revertedWithPanic", function (expectedCodeArg) {
        var _a;
        const ethers = require("ethers");
        let expectedCode;
        try {
            if (expectedCodeArg !== undefined) {
                const normalizedCode = (0, common_1.normalizeToBigInt)(expectedCodeArg);
                expectedCode = ethers.BigNumber.from(normalizedCode);
            }
        }
        catch (_b) {
            throw new TypeError(`Expected the given panic code to be a number-like value, but got '${expectedCodeArg}'`);
        }
        const code = expectedCode;
        let description;
        let formattedPanicCode;
        if (code === undefined) {
            formattedPanicCode = "some panic code";
        }
        else {
            const codeBN = ethers.BigNumber.from(code);
            description = (_a = (0, panic_1.panicErrorCodeToReason)(codeBN)) !== null && _a !== void 0 ? _a : "unknown panic code";
            formattedPanicCode = `panic code ${codeBN.toHexString()} (${description})`;
        }
        const onSuccess = () => {
            this.assert(false, `Expected transaction to be reverted with ${formattedPanicCode}, but it didn't revert`);
        };
        const onError = (error) => {
            const returnData = (0, utils_1.getReturnDataFromError)(error);
            const decodedReturnData = (0, utils_1.decodeReturnData)(returnData);
            if (decodedReturnData.kind === "Empty") {
                this.assert(false, `Expected transaction to be reverted with ${formattedPanicCode}, but it reverted without a reason`);
            }
            else if (decodedReturnData.kind === "Error") {
                this.assert(false, `Expected transaction to be reverted with ${formattedPanicCode}, but it reverted with reason '${decodedReturnData.reason}'`);
            }
            else if (decodedReturnData.kind === "Panic") {
                if (code !== undefined) {
                    this.assert(decodedReturnData.code.eq(code), `Expected transaction to be reverted with ${formattedPanicCode}, but it reverted with panic code ${decodedReturnData.code.toHexString()} (${decodedReturnData.description})`, `Expected transaction NOT to be reverted with ${formattedPanicCode}, but it was`);
                }
                else {
                    this.assert(true, null, `Expected transaction NOT to be reverted with ${formattedPanicCode}, but it reverted with panic code ${decodedReturnData.code.toHexString()} (${decodedReturnData.description})`);
                }
            }
            else if (decodedReturnData.kind === "Custom") {
                this.assert(false, `Expected transaction to be reverted with ${formattedPanicCode}, but it reverted with a custom error`);
            }
            else {
                const _exhaustiveCheck = decodedReturnData;
            }
        };
        const derivedPromise = Promise.resolve(this._obj).then(onSuccess, onError);
        this.then = derivedPromise.then.bind(derivedPromise);
        this.catch = derivedPromise.catch.bind(derivedPromise);
        return this;
    });
}
exports.supportRevertedWithPanic = supportRevertedWithPanic;
//# sourceMappingURL=revertedWithPanic.js.map